<b>Привет!</b>

Меня зовут Харочкина Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовила замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> 
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> 
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Таким комментарием выделены удачные решения🙂
</div>


Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

3.1 Чтобы найти зависимость между количеством детей и возвратом кредита в срок или ее отсутствие, создадим сводную таблицу со следующими столбцами:

*children* - количество детей в семье клиента;

*debt* - имел ли клиент задолженность по возврату кредитов;

*count* - общее количество записей в данной категории (количество детей);

*percent* - процент задолженности в категории.

In [22]:
import numpy as np
data_children=data.pivot_table(index='children', values=['family_status', 'debt'],
                               aggfunc={'family_status': len,
                                        'debt': np.sum})
data_children = data_children.rename(columns={'family_status':'count'})
data_children['percent'] = data_children['debt']/data_children['count']*100
display(data_children.sort_values(by='percent', ascending=False))
all_count=data_children.loc[1:, 'count'].sum()
all_debt=data_children.loc[1:, 'debt'].sum()
print('Общий процент задолженности клиентов, не имеющих детей: '+ str((data_children.loc[0, 'debt']/data_children.loc[0, 'count'])*100)[:4]+'%.')
print('Общий процент задолженности клиентов, имеющих детей: '+ str((all_debt/all_count)*100)[:4]+'%.')

debt  count   percent
children                       
4            4     41  9.756098
2          194   2052  9.454191
1          444   4808  9.234609
3           27    330  8.181818
0         1063  14091  7.543822
5            0      9  0.000000

Общий процент задолженности клиентов, не имеющих детей: 7.54%.
Общий процент задолженности клиентов, имеющих детей: 9.24%.


<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> 
    
Все библиотеки нужно импортировать в одной ячейке в начале всего кода, это соответствует стилистическому гайду <a href="https://peps.python.org/pep-0008/">pep 8</a>
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Хорошо, что используешь `pivot_table()` Чтобы найти долю должников, нужно количество должников поделить на общее число заемщиков. Эту долю можно найти с помощью функции агрегации `mean` Для округления значений можно использовать `round()` 
    
Также могу посоветовать переименовать столбцы и написать функцию для построения сводных во всех вопросах.
</div>

**Вывод:** 

В общем, клиенты, имеющие детей, имеют задолженность по возврату кредитов в 9 из 100 случаев. С клиентами, не имеющими детей, такое случается уже в 7 из 100.

Клиенты, имеющие до 4 детей, имеют больший процент задолженности, чем клиенты без детей. Значит, зависимость существует. Напротив клиенты, имеющие 5 детей, кажутся самыми дисциплинированными. Однако, записей про них совсем немного, а потому утверждать об их повышенной благонадежности нельзя. Клиентов, имеющих 4 детей, также немного, однако процент их задолженности похож на процент задолженности других клиентов с детьми, поэтому можно считать это условно верным.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Хороший вывод
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

3.2 Чтобы найти зависимость между семейным положением и возвратом кредита в срок или ее отсутствие, создадим сводную таблицу со следующими столбцами:

*family_status* - семейный статус;

*debt* - имел ли клиент задолженность по возврату кредитов;

*count* - общее количество записей в данной категории (семейный статус);

*percent* - процент задолженности в категории.

In [23]:
data_fam=data.pivot_table(index='family_status', values=['debt', 'children'],
                               aggfunc={'children': len,
                                        'debt': np.sum})
data_fam = data_fam.rename(columns={'children':'count'})
data_fam['percent'] = data_fam['debt']/data_fam['count']*100
display(data_fam.sort_values(by='percent', ascending=False))

count  debt   percent
family_status                               
Не женат / не замужем   2796   273  9.763948
гражданский брак        4134   385  9.313014
женат / замужем        12261   927  7.560558
в разводе               1189    84  7.064760
вдовец / вдова           951    63  6.624606

**Вывод:** 

Клиенты, не заключавшие и не находящиеся в официальном браке, чаще имеют проблемы с возвратом задолженности (9,76% и 9,31%). Клиенты, состоявшие и находящиеся в браке, имеют меньше просроченнной задолженности (7,56%, 7,06%, 6,62%). Это также указывает на наличие зависимости семейного положения и наличия задолженности по выплатам.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Хорошо
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Используем ранее введенные категории уровня дохода:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

In [24]:
data_inc=data.pivot_table(index='total_income_category', values=['debt', 'children'],
                               aggfunc={'children': len,
                                        'debt': np.sum})
data_inc = data_inc.rename(columns={'children':'count'})
data_inc['percent'] = data_inc['debt']/data_inc['count']*100
display(data_inc.sort_values(by='percent', ascending=False))
print(len(data))

count  debt   percent
total_income_category                       
E                         22     2  9.090909
C                      15921  1353  8.498210
A                         25     2  8.000000
B                       5014   354  7.060231
D                        349    21  6.017192

21331


**Вывод:** 

Данная категоризация слишком неравно делит количество всех имеющихся записей, а потому наверняка утверждать что-то о корректности выводов, допустим, о категории "Е" нельзя. Казалось бы, клиенты, получающие меньше остальных (до 30 тыс. рублей, категория Е), выплачивают кредиты хуже остальных, но и берут кредиты они нечасто. Всего 22 случая из 21 тысячи. Проблемы с выплатами у этой категории звучат вполне логично, но выборка все таки слишком мала, чтобы считать ее репрезентативной. Если опустить эту категорию, то категория "С" имеет проблемы с выплатами чаще остальных. Однако, использовать результаты данного анализа при построении модели кредитного скоринга не стоит, ведь следующая категория "D" кажется самой благонадежной. Это выглядит неправдоподобно. Да и, в принципе, среди этих категорий сложно проследить зависимость. Возможно, если переработать категоризацию дохода, зависимость найдется.

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> 
    
Здесь лучше делать выводы по двум самым крупным категориям - B и С
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_pur=data.pivot_table(index='purpose_category', values=['debt', 'children'],
                               aggfunc={'children': len,
                                        'debt': np.sum})
data_pur = data_pur.rename(columns={'children':'count'})
data_pur['percent'] = data_pur['debt']/data_pur['count']*100
display(data_pur.sort_values(by='percent', ascending=False))

count  debt   percent
purpose_category                               
операции с автомобилем     4279   400  9.347978
получение образования      3988   369  9.252758
проведение свадьбы         2313   183  7.911803
операции с недвижимостью  10751   780  7.255139

**Вывод:** 

Клиенты, оформляющие кредит для совешения операций с операций с автомобилем или получения образования, имеют больше проблем с выплатой задолженности (9,35% и 9,25% соотвественно). Напротив, клиенты, оформляющие кредит для операций с недвижимостью и организации свадьбы, возвращают задолженность в срок чаще (7,26% и 7,91% соответственно). А, значит, зависимость существует.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Верно
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Изначально пропуски обнаружились в столбцах "days_employed" и "total_income". Первая причина, которая приходит на ум-это отсуствие у клиента работы, а, вследствие, и общего трудового стажа, и общего дохода. Однако, такая теория применима в основном к студентам, но пропуски встречаются как у студентов, так и у сотрудников, пенсионеров, компаньонов. Значит, причины могли оказаться техническими или человеческими. Либо сбой в программе, либо сотрудник банка/клиент не ввел эти данные. Кроме того, возможно, это просто ошибки выгрузки. 

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Хорошие предположения. Возможно, эту информацию не требуют в некоторых случаях.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Среднее значение далеко не всегда точно описывает ситуацию, а вот медиана устойчива к аномальным отклонениям (выбросам). Как в известном анекдоте: у Пети 10 яблок, а у Васи 0 — в среднем у обоих по 5 яблок. Когда в выборке присутствуют выдающиеся значения (300 против 100), лучше использовать медиану.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Да, смотрим на наличие выбросов👍
</div>

### Шаг 4: общий вывод.

При построении модели кредитного скоринга (специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку) стоит учитывать такие факторы как:

1. Количество детей кредитополучателя (процент задолженности клиентов, не имеющих детей: 7.54%; процент задолженности клиентов, имеющих детей: 9.24%);

2. Семейное положение клиента (клиенты, не заключавшие и не находящиеся в официальном браке, чаще имеют проблемы с возвратом задолженности (9,76% и 9,31%);

3. Цель кредита (клиенты, оформляющие кредит для совешения операций с операций с автомобилем или получения образования, имеют больше проблем с выплатой задолженности (9,35% и 9,25% соотвественно)).

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> 
    
Также можно добавить рекомендации банку-заказчику, описать портрет "идеального" клиента и/или, наоборот, наиболее рискованного клиента.
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
Отлично! Критических замечаний нет. Проект принят! 👏 Поздравляю и желаю удачи!
</font></b>
</div>